### ISSUE TITLE
May I have ARM 64 version?

### ISSUE THREAD

**Issue by raychan907** [thijsvanloef/palworld-server-docker#10]
I tried your docker image but not work. It shows that exec format error. I guess my Linux is ARM64. So, May I get the ARM version for this docker image? Thank you very much.

**Comment by wardlycode** [comment]
Would be huge for running in my pi cluster :D

**Comment by raychan907** [comment]
![Palworld_error_log](https://github.com/thijsvanloef/palworld-server-docker/assets/35477166/6e361844-10ba-4ce4-afbc-c7bd82e692b1)

Herer is the log, It seems not work.
Many Thanks

**Comment by alex4108** [comment]
There's an old example I can provide for using Docker BuildX to compile arm64 format container using x86 host: https://github.com/alex4108/Approova/blob/main/.github/workflows/release.yml

Latest seems we can use build-push as buildx is included: https://github.com/docker/build-push-action/

I might submit a PR for this soon :D

**Comment by thijsvanloef** [comment]
Hi! The steamcmd docker image this image is based on, does not support ARM64. Running this container on ARM devices (Raspberry PI, Oracle Ampere A1) would require an overhaul of the Dockerfile.

**Comment by paulwer** [comment]
maybe this could be used? https://hub.docker.com/r/teriyakigod/steamcmd
I am not that good into docker image build process, so srry it this is unnecessary :(

**Comment by JohnC0de** [comment]
Same issue here, if somebody could make an arm64 version please

**Comment by BZX6666** [comment]
maybe this could be used https://www.nodeseek.com/post-61273-1

**Comment by Berodin** [comment]
> Would be huge for running in my pi cluster :D

perhaps a kubernetes cluster build on pi? :D

**Comment by alex4108** [comment]
I don't have any time left today and don't know when I'll be able to pick back up, but there's a WIP here, if someone else picks up feel free to base off this.  I might be able to come back tomorrow and see what else I can do.

https://github.com/alex4108/palworld-server-docker
* Using azure d2pls_v5 vm (ubuntu 22)
* clone/fork the repo to something with aarch64
* `docker-compose kill && docker-compose rm -f && git pull && make build-arm && docker-compose up -d` to run the latest code

There are some `TODO`'s sprinkled in here about where I'm at.  If someone else picks up in the meantime, godspeed and please post here so we don't duplicate efforts :D

it looks like init.sh will complete successfully now, but I didn't let it run all the way yet.  I just now got to a point where steamcmd runs.

**Comment by thijsvanloef** [comment]
@alex4108 Thank you for your work so far, feel free to open a draft pull request if you are looking for input :)

**Comment by yumusb** [comment]
https://github.com/yumusb/palworld-server-docker-arm 
I added the arm64 version to this project with the dockerfile from https://github.com/TeriyakiGod/steamcmd-docker-arm64

**Comment by BaileyMillerSSI** [comment]
Would an arm version even work because the server executable would need to support arm?

**Comment by yumusb** [comment]
> Would an arm version even work because the server executable would need to support arm?

Currently steamcmd doesn't support arm, it's running on https://github.com/FEX-Emu/FEX right now.

**Comment by lichuu** [comment]
```
diff --git a/scripts/init.sh b/scripts/init.sh
index 7f507f3..2147912 100644
--- a/scripts/init.sh
+++ b/scripts/init.sh
@@ -33,7 +33,7 @@ if [ "${UPDATE_ON_BOOT}" = true ]; then
 printf "\e[0;32m*****STARTING INSTALL/UPDATE*****\e[0m\n"
     if [ "$is_arm" == "true" ]; then # if this an ARM kernel "aarch64", use FEXBash
         FEXBash="FEXBash"
-        ls -latr /home/steam/Steam/
+        ls -latR /home/steam/Steam/
         steamcmd="/home/steam/Steam/steamcmd.sh"
     else
         FEXBash=""
@@ -43,6 +43,9 @@ printf "\e[0;32m*****STARTING INSTALL/UPDATE*****\e[0m\n"
     chmod +x $steamcmd

     su $USERNAME -c "${FEXBash} -c \"${steamcmd} +force_install_dir /palworld +login anonymous +app_update 2394010 validate +quit\""
+    su $USERNAME -c "${FEXBash} -c \"${steamcmd} +login anonymous +app_update 1007 validate +quit\""
+    mkdir -p /home/steam/.steam/sdk64/
+    cp /home/steam/Steam/steamapps/common/Steamworks\ SDK\ Redist/linux64/steamclient.so /home/steam/.steam/sdk64/
 fi


diff --git a/scripts/start.sh b/scripts/start.sh
index 2be1183..0a864a3 100644
--- a/scripts/start.sh
+++ b/scripts/start.sh
@@ -2,7 +2,7 @@

 # TODO Haven't tested this at all....

-STARTCOMMAND="./PalServer.sh"
+STARTCOMMAND="FEXBash ./PalServer.sh"

 if [ -n "${PORT}" ]; then
     STARTCOMMAND="${STARTCOMMAND} -port=${PORT}"
```

This with @alex4108's branch got it working. definitely could use some more love though...

**Comment by zerobell-lee** [comment]
@yumusb 

1. Need to fix docker-compose.yml

```yaml
services:
  palworld:
    build:
      context: .
      dockerfile: Dockerfile.arm
```

or better specify version of yaml

2. error occurs on building

```bash
Step 10/22 : RUN mkdir Ubuntu &&     cd Ubuntu &&     curl -4 'https://github.com/TeriyakiGod/steamcmd-docker-arm64/raw/a1861e970a2e956f9471bedc57055278d235f422/Ubuntu_22_04.tar.gz' -Lo Ubuntu.tar.gz &&     tar xzf Ubuntu.tar.gz &&     rm ./Ubuntu.tar.gz &&     cd ../../ &&     echo '{"Config":{"RootFS":"Ubuntu/Ubuntu_22_04"}}' > ./Config.json
 ---> Running in 1f55f83284b7
mkdir: cannot create directory 'Ubuntu': Permission denied
```

I'm using Ubuntu 22 & OCI & A1.flex
Maybe I need to put `sudo` or somethign? But I don't know why it doesn't work if it worked on your machine.

**Comment by yumusb** [comment]
> @yumusb
> 
>     1. Need to fix docker-compose.yml
> 
> 
> ```yaml
> services:
>   palworld:
>     build:
>       context: .
>       dockerfile: Dockerfile.arm
> ```
> 
> or better specify version of yaml
> 
>     2. error occurs on building
> 
> 
> ```shell
> Step 10/22 : RUN mkdir Ubuntu &&     cd Ubuntu &&     curl -4 'https://github.com/TeriyakiGod/steamcmd-docker-arm64/raw/a1861e970a2e956f9471bedc57055278d235f422/Ubuntu_22_04.tar.gz' -Lo Ubuntu.tar.gz &&     tar xzf Ubuntu.tar.gz &&     rm ./Ubuntu.tar.gz &&     cd ../../ &&     echo '{"Config":{"RootFS":"Ubuntu/Ubuntu_22_04"}}' > ./Config.json
>  ---> Running in 1f55f83284b7
> mkdir: cannot create directory 'Ubuntu': Permission denied
> ```
> 
> I'm using Ubuntu 22 & OCI & A1.flex Maybe I need to put `sudo` or somethign? But I don't know why it doesn't work if it worked on your machine.



test on this.
<img width="1707" alt="图片" src="https://github.com/thijsvanloef/palworld-server-docker/assets/43062104/06afb160-6b04-4a7c-945f-2d8b2cc26333">

**Comment by yumusb** [comment]
> @yumusb
> 
>     1. Need to fix docker-compose.yml
> 
> 
> ```yaml
> services:
>   palworld:
>     build:
>       context: .
>       dockerfile: Dockerfile.arm
> ```
> 
> or better specify version of yaml
> 
>     2. error occurs on building
> 
> 
> ```shell
> Step 10/22 : RUN mkdir Ubuntu &&     cd Ubuntu &&     curl -4 'https://github.com/TeriyakiGod/steamcmd-docker-arm64/raw/a1861e970a2e956f9471bedc57055278d235f422/Ubuntu_22_04.tar.gz' -Lo Ubuntu.tar.gz &&     tar xzf Ubuntu.tar.gz &&     rm ./Ubuntu.tar.gz &&     cd ../../ &&     echo '{"Config":{"RootFS":"Ubuntu/Ubuntu_22_04"}}' > ./Config.json
>  ---> Running in 1f55f83284b7
> mkdir: cannot create directory 'Ubuntu': Permission denied
> ```
> 
> I'm using Ubuntu 22 & OCI & A1.flex Maybe I need to put `sudo` or somethign? But I don't know why it doesn't work if it worked on your machine.

@zerobell-lee 

I'm using Ubuntu 22 & OCI & A1.flex, too. Please use `root` user to try. 
<img width="1359" alt="图片" src="https://github.com/thijsvanloef/palworld-server-docker/assets/43062104/1d00f8d7-69af-45f2-bccc-63dac0b6bcb2">

And fixed healthcheck command. Please pull and try again.

**Comment by mmouradev** [comment]
@yumusb 
I'm having this error, missing something?
`palworld-1  | *****EXECUTING USERMOD*****
palworld-1  | usermod: no changes
palworld-1  | *****EXECUTING USERMOD COMPLETE*****
palworld-1  | *****STARTING INSTALL/UPDATE*****
palworld-1  | arm architecture detected, using FEXBash to run steamcmd
palworld-1  | Ubuntu/Ubuntu_22_04/bin/sh: command not found
palworld-1  | *****INSTALL/UPDATE COMPLETE*****
palworld-1  | ln: failed to create symbolic link '/home/steam/.steam/sdk64/steamclient.so': File exists
palworld-1  | *****CHECKING FOR EXISTING CONFIG*****
palworld-1  | *****GENERATING CONFIG*****
palworld-1  | Ubuntu/Ubuntu_22_04/bin/sh: command not found
palworld-1  | cp: cannot stat '/palworld/DefaultPalWorldSettings.ini': No such file or directory
palworld-1  | RCON_ENABLED=true
palworld-1  | sed: can't read /palworld/Pal/Saved/Config/LinuxServer/PalWorldSettings.ini: No such file or directory
palworld-1  | RCON_PORT=25575
palworld-1  | sed: can't read /palworld/Pal/Saved/Config/LinuxServer/PalWorldSettings.ini: No such file or directory
palworld-1  | *****STARTING SERVER*****
palworld-1  | ./PalServer.sh -port=8211 -players=16 -adminpassword="adminPasswordHere" -queryport=27015 -useperfthreads -NoAsyncLoadingThread -UseMultithreadForDS
palworld-1  | arm architecture detected, using FEXBash to start server
palworld-1  | Ubuntu/Ubuntu_22_04/bin/sh: command not found`

**Comment by yumusb** [comment]
https://t.me/+836Dek4SfYA4NDQ9  

@mmouradev @zerobell-lee  

Please add this tg group chat and I will help you guys debug.




Anyone else having problems with ARM can join this group chat.

**Comment by minjae9610** [comment]
![image](https://github.com/thijsvanloef/palworld-server-docker/assets/33440010/4962d83d-91e5-435d-a943-1cda4a75f2b8)
I hope it works on my raspberry pi 5, I'm trying to figure out how too.

**Comment by funkymango221** [comment]
Server was working fine on arm with a fresh save before I transferred my old world save from windows server to oracle's free tier arm64. These are the steps I followed to do so:
1. Download Pal/Saved/SaveGames from windows server you are trying to transfer over

Make sure your Linux Palword server is a clean install and not something you experimented on

2. Start Linux server and let boot to generate files.

3. Stop Linux Server (Do not start the server again until the end)

4. Delete the Pal/Saved/SaveGames from the Linux server, and replace it with the one from Windows. It should now look like Pal/Saved/SaveGames/0/(bunch of random numbers)

5. In "/Pal/Saved/Config/LinuxServer" edit "GameUserSettings.ini" and change "DedicatedServerName" to the random numbers in your "Pal/Saved/SaveGames/0/" folder (example - DedicatedServerName=77ESD46EFEDFGSDF5559CD1EB6D4F)

Most people have already done above, next steps are super important

6. In linux terminal navigate to /Pal/Binaries/Linux (you should see files like PalServer-Linux)

Enter the following commands in terminal:

echo "1623730" > steam_appid.txt
chmod 444 steam_appid.txt
chattr +i steam_appid.txt   (this stops the pal linux binary from changing the appid and deleting, if you ever want to delete this file you need to use -i flag instead)

7. Check the contents of your steam_appid.txt file. It should say 1623730

8. Now start your linux server and let it fully boot

9. Check the contents of your steam_appid.txt file again, it should still say 1623730. If it doesn't you might have missed one of the commands, delete the "steam_appid.txt" file and try step 6 again.

10. Login to the server, all your player data should now be there as normal

Now when I do "docker compose up -d", I receive this message.
![image](https://github.com/thijsvanloef/palworld-server-docker/assets/157784755/a0856742-b9e1-4b14-ad71-c98f022bcb5a)

**Comment by jordanmckoy** [comment]
> hen I do "docker compose up -d", I receive this message.
Hey did you ever get this solved? I am trying to run this also on an oracle free tier vm

**Comment by RhanCandia** [comment]
> Server was working fine on arm with a fresh save before I transferred my old world save from windows server to oracle's free tier arm64. These are the steps I followed to do so:
> 
> 1. Download Pal/Saved/SaveGames from windows server you are trying to transfer over
> 
> Make sure your Linux Palword server is a clean install and not something you experimented on
> 
> 2. Start Linux server and let boot to generate files.
> 3. Stop Linux Server (Do not start the server again until the end)
> 4. Delete the Pal/Saved/SaveGames from the Linux server, and replace it with the one from Windows. It should now look like Pal/Saved/SaveGames/0/(bunch of random numbers)
> 5. In "/Pal/Saved/Config/LinuxServer" edit "GameUserSettings.ini" and change "DedicatedServerName" to the random numbers in your "Pal/Saved/SaveGames/0/" folder (example - DedicatedServerName=77ESD46EFEDFGSDF5559CD1EB6D4F)
> 
> Most people have already done above, next steps are super important
> 
> 6. In linux terminal navigate to /Pal/Binaries/Linux (you should see files like PalServer-Linux)
> 
> Enter the following commands in terminal:
> 
> echo "1623730" > steam_appid.txt chmod 444 steam_appid.txt chattr +i steam_appid.txt (this stops the pal linux binary from changing the appid and deleting, if you ever want to delete this file you need to use -i flag instead)
> 
> 7. Check the contents of your steam_appid.txt file. It should say 1623730
> 8. Now start your linux server and let it fully boot
> 9. Check the contents of your steam_appid.txt file again, it should still say 1623730. If it doesn't you might have missed one of the commands, delete the "steam_appid.txt" file and try step 6 again.
> 10. Login to the server, all your player data should now be there as normal
> 
> Now when I do "docker compose up -d", I receive this message. ![image](https://private-user-images.githubusercontent.com/157784755/300079038-a0856742-b9e1-4b14-ad71-c98f022bcb5a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY1OTMxODAsIm5iZiI6MTcwNjU5Mjg4MCwicGF0aCI6Ii8xNTc3ODQ3NTUvMzAwMDc5MDM4LWEwODU2NzQyLWI5ZTEtNGIxNC1hZDcxLWM5OGYwMjJiY2I1YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEzMFQwNTM0NDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xOWU5MTUwMzM2NzY1NThjMWUyMjMxYzM0NTkwNmYxYWNkZTRhZjMxYzQ2ZjgzYzhmMThlZmQyMDU1MWEyZjgwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.e7Hbk5cu-B2I2yUzkYFioe00dBSaujIZx7IW0XISXYU)

Hi there. After you transfer the old world save files into the appropriate directory, have you tried running a `chmod -R 777 <savefile directory>`? Maybe the game is failing the autosave because it doesn't have the right permissions on the old world's save files.

**Comment by thijsvanloef** [comment]
New release has been deployed, you can now use the `:latest-arm64` tag to deploy your container on ARM64 devices :)

**Comment by RhanCandia** [comment]
> New release has been deployed, you can now use the `:latest-arm64` tag to deploy your container on ARM64 devices :)

That's awesome! Thank you. I'll try it out.

**Comment by raychan907** [comment]
> New release has been deployed, you can now use the `:latest-arm64` tag to deploy your container on ARM64 devices :)

Awesome!! Thank you very much for your help.
Cheers.
